### ISSUE TITLE
MenSelmapPos_TopN__{0,1} from Brawl cannot be edited properly

### ISSUE THREAD

**Issue by libertyernie** [libertyernie/brawltools#45]
To reproduce, go to MenSelmapPos, animation MenSelmapPos_TopN__0 (in common5_en.pac from Brawl) and try editing the X translation of bone pos21 on frame 42. BrawlBox shows -3.15586 as the X translation for frame 42 no matter what you do, but the translation you try to enter for frame 42 is reflected on frame 43.

Also, when opening the original Brawl version of the file, each frame N corresponds to the screen with N-3 stage icons. When opening a file edited with an earlier version of BrawlBox (and which works properly in Brawl), frame N corresponds to N-2 stage icons.

**Comment by IsaacSchemm** [comment]
This code in KeyframeCollection causes the problem:
<pre>
                //There can be up to two keyframes with the same index.
                if (!parsing || OLD_KEYFRAME_PARSING)
                    entry._value = value; //Do this when editing
                else
                {
                    //And this when parsing
                    _keyCount++;
                    KeyframeEntry temp = new KeyframeEntry(index, value);
                    entry.InsertAfter(temp);
                    entry = temp;
                }
</pre>
That second block is new (added in December), and it seems to expose some very odd parts of Nintendo's original MenSelmapPos_TopN__0. Why would it have two keyframes on the same index?

Here's a printout of just the Y translation for just frame 34. When there are two, I think Brawl is just using the second one. Of course, this is the SSS and not a stage or character, so it might be different.
<pre>MenSelmapPos_TopN
pos00
pos01
16.69126
pos02
16.69126
pos03
16.69126
pos04
16.69126
pos05
16.69126
pos06
10.89507
pos07
10.89268
pos08
10.89268
pos09
10.89268
pos10
10.89268
pos11
10.89268
pos12
5.097803
pos13
5.097803
pos14
5.094091
pos15
5.094091
pos16
5.094091
pos17
-0.7032719
5.094091
pos18
-0.7032719
pos19
-0.7032719
pos20
-0.7032719
pos21
-0.7044926
pos22
-6.50083
-0.7044926
pos23
-6.50083
pos24
-6.50083
pos25
-6.50083
pos26
-6.50083
pos27
-6.50083
pos28
-12.30308
-6.503077
pos29
-12.30308
pos30
-12.30308
pos31
-12.30308
pos32
-19.1
-12.30308
pos33
-11.00154
-19.1
pos34
-11.00154
pos35
-11.00198
pos36
-11.00198
pos37
-11.00198
pos38
-11.00198
pos39
-11.00198
pos40
-11.00198
pos41
</pre>

**Comment by BlackJax96** [comment]
Two keyframes (and only two) can have the same frame index because the first one in memory is the in keyframe, and the second in memory is the out keyframe. Game&Watch uses two keyframes on one frame in his animations to jump from one pose to the next.

The CHR0 editor was not made with editing two keyframes on one frame in mind because it was a fairly recent realization (g&w couldn't be animated properly before). It really needs to be updated.

Currently you can only edit it properly with the interpolation editor (in the model viewer, Tools->Interpolation Editor), which will show the information of both keyframes on that frame.

**Comment by libertyernie** [comment]
OK, thanks. This came about when I was working on the preview function for SSS Editor, so I'll find a way to access the out keyframe directly when there are two.

**Comment by libertyernie** [comment]
Does this make sense to put in CHR0Node? Or are there any improvements you can suggest?
It's a modified version of GetAnimFrame that looks for other keyframes at the same index, and uses the keyframe value if there is one instead of the GetFrameValue result.

        public CHRAnimationFrame GetAnimOutFrame(int index)
        {
            CHRAnimationFrame frame = new CHRAnimationFrame() { Index = index };
            float* dPtr = (float*)&frame;
            for (int x = 0; x < 9; x++)
            {
                KeyframeEntry keyframe = Keyframes.GetKeyframe(x, index);
                while (keyframe != null && keyframe._next != null && keyframe._next._index == keyframe._index)
                    keyframe = keyframe._next;
                frame.SetBool(x, keyframe != null);
                *dPtr++ = keyframe != null
                    ? keyframe._value
                    : GetFrameValue(x, index);
            }
            return frame;
        }

**Comment by BlackJax96** [comment]
Yeah, that looks good! I do have a suggestion to make it easier to do the same thing in other animation types later on though.

In KeyframeArray, you could add "bool returnOutValue = false" as an argument to "GetFrameValue" and inside the 'if' where it says "if (entry._index == index) //The index is a keyframe" you could do the search for the out keyframe there instead if returnOutValue is true.

Then you can just call GetFrameValue(x, index, true) and it will return the out value if it exists. You'll have to update GetFrameValue in CHR0Node and in KeyframeCollection to handle the new bool as well.

**Comment by PeterHatch** [comment]
Wanted to note that I'm seeing what is probably this same problem (in 0.76b) with MenSelchrPos, and similar animations.  I've just been editing in 0.73b for now...

**Comment by libertyernie** [comment]
Yeah, I added that option on the BrawlLib code, but I never allowed the model editor in BrawlBox to actually use it to edit the animations - wasn't quite sure what the interface for that would be like. I just wanted to make sure SSS Editor would show the correct thing.
I suppose if I have to edit that file again I'll just use an old BrawlBox too (like BJ said, those work because the second keyframe is discarded and so Brawl just sees the one), or figure out the interpolation editor and see if I can use that.
